File: service/research/training/service.py
from __future__ import annotations

from copy import deepcopy
from dataclasses import dataclass

from service.research.training import orchestrator
from service.research.training.presenter import build_training_session_payload
from service.research.training.session_store import TrainingSessionStore


@dataclass(frozen=True)
class TrainingConfig:
    ruleset: str = "default"
    bot_count: int = 5
    generations: int = 1
    base_genome_id: str = "random"
    bot_model: str = "genetic"
    mutation_strength: float = 0.25
    mutation_rate: float = 0.15
    random_immigrant_count: int = 1
    games_per_generation: int = 5

    def __post_init__(self):
        object.__setattr__(self, "games_per_generation", max(1, min(50, int(self.games_per_generation))))


class TrainingService:
    def __init__(self, database, game_factory, store=None,
                 bot_client_factory=None):
        self.database = database
        self.store = store or TrainingSessionStore()
        self.runtime = orchestrator.TrainingRuntime(
            database=database,
            game_factory=game_factory,
            bot_client_factory=bot_client_factory,
            sessions=self.store._runtime_sessions(),
        )

    async def start(self, config: TrainingConfig):
        return await orchestrator.start_training_session(
            self.runtime,
            config.ruleset, config.bot_count, config.generations,
            config.base_genome_id, config.bot_model,
            mutation_strength=config.mutation_strength,
            mutation_rate=config.mutation_rate,
            random_immigrant_count=config.random_immigrant_count,
            games_per_generation=config.games_per_generation,
        )

    async def cancel(self, session_id: str,
                     reason: str = "Training cancelled by operator") -> bool:
        return await orchestrator.cancel_training_session(
            self.runtime, session_id, reason)

    async def rerun(self, batch_id: str):
        batch = self.database.get_training_batch(batch_id)
        if not batch:
            return None
        options = batch.get("config", {}) or {}
        return await self.start(TrainingConfig(
            ruleset=batch.get("ruleset", "default"),
            bot_count=int(batch.get("bot_count") or 5),
            generations=int(batch.get("total_generations") or 1),
            base_genome_id=batch.get("base_genome_id") or "random",
            bot_model=batch.get("bot_model", "genetic"),
            mutation_strength=float(options.get("mutation_strength", 0.25)),
            mutation_rate=float(options.get("mutation_rate", 0.15)),
            random_immigrant_count=int(options.get("random_immigrant_count", 1)),
            games_per_generation=int(options.get("games_per_generation", 5)),
        ))

    def list(self):
        return build_training_session_payload(self.store.list())

    def status(self, session_id: str):
        return deepcopy(self.store.get(session_id))

    async def handle_game_ended(self, game_id: str, session_id: str):
        return await orchestrator.handle_training_game_ended(
            self.runtime, game_id, session_id)

    async def watchdog_loop(self, interval_seconds: int = 30,
                            stale_after_seconds: int = 600):
        return await orchestrator.training_watchdog_loop(
            self.runtime, interval_seconds, stale_after_seconds)

File: service/src/research/training/service.ts
import type { TrainingSessionsDTO } from "@takes-a-village/shared";

import type { DatabaseProvider } from "../../db/contracts.js";
import type { CreateGameOptions } from "../../game-manager/lifecycle.js";
import type { TrainingBotClient } from "./bot-client.js";
import { createTrainingConfig, type TrainingConfigInput } from "./config.js";
import { TrainingOrchestrator, type TrainingUpdatePublisher } from "./orchestrator.js";
import { presentTrainingSessions } from "./presenter.js";
import { TrainingSessionStore, type TrainingSession } from "./session-store.js";

export interface TrainingServiceOptions {
  database: DatabaseProvider;
  createGame(hostId: string, ruleset: string, options: CreateGameOptions): string;
  botClient: TrainingBotClient;
  updateHub: TrainingUpdatePublisher;
  store?: TrainingSessionStore;
  idFactory: () => string;
  random?: () => number;
  clock?: () => Date;
}

function numericOption(value: unknown, fallback: number): number { return typeof value === "number" ? value : fallback; }

export class TrainingService {
  readonly store: TrainingSessionStore;
  private readonly orchestrator: TrainingOrchestrator;

  constructor(private readonly options: TrainingServiceOptions) {
    this.store = options.store ?? new TrainingSessionStore();
    this.orchestrator = new TrainingOrchestrator({
      ...options,
      sessions: this.store.runtimeSessions(),
    });
  }

  async start(input: TrainingConfigInput = {}): Promise<string> {
    return this.orchestrator.start(createTrainingConfig(input));
  }

  async cancel(sessionId: string, reason = "Training cancelled by operator"): Promise<boolean> {
    return this.orchestrator.cancel(sessionId, reason);
  }

  async rerun(batchId: string): Promise<string | null> {
    const batch = await this.options.database.getTrainingBatch(batchId);
    if (!batch) return null;
    const config = batch.config ?? {};
    return this.start({
      ruleset: batch.ruleset ?? "default",
      botCount: batch.bot_count ?? 5,
      generations: batch.total_generations ?? 1,
      baseGenomeId: batch.base_genome_id ?? "random",
      botModel: batch.bot_model ?? "genetic",
      mutationStrength: numericOption(config.mutation_strength, 0.25),
      mutationRate: numericOption(config.mutation_rate, 0.15),
      randomImmigrantCount: numericOption(config.random_immigrant_count, 1),
      gamesPerGeneration: numericOption(config.games_per_generation, batch.games_per_generation ?? 5),
    });
  }

  list(): TrainingSessionsDTO { return presentTrainingSessions(this.store.runtimeSessions()); }
  status(sessionId: string): TrainingSession | undefined { return this.store.get(sessionId); }
  async handleGameEnded(gameId: string, sessionId: string): Promise<void> { await this.orchestrator.handleGameEnded(gameId, sessionId); }
  async reconcileStalled(staleAfterMilliseconds?: number, attemptStaleAfterMilliseconds?: number): Promise<void> {
    await this.orchestrator.reconcileStalled(staleAfterMilliseconds, attemptStaleAfterMilliseconds);
  }
}
